File: nz-legislation-tool/src/commands/get.ts
/**
 * Get command - Retrieve legislation by ID
 */

import { Command } from 'commander';
import ora from 'ora';

import { getWork, getWorkVersions } from '@client';
import { printWorkDetail, printVersionsTable, printJson, versionsToCsv } from '@output';

interface GetOptions {
  versions?: boolean;
  format: string;
}

export const getCommand = new Command()
  .name('get')
  .description('Get legislation by ID')
  .argument('<id>', 'Work ID (e.g., act/2020/67)')
  .option('--versions', 'Show version history')
  .option('--format <format>', 'Output format (table, json, csv)', 'table')
  .action(async (workId: string, options: GetOptions) => {
    const spinner = ora('Retrieving legislation...').start();

    try {
      if (options.versions) {
        // Get version history
        const versions = await getWorkVersions(workId);
        spinner.stop();

        switch (options.format.toLowerCase()) {
          case 'json':
            printJson(versions);
            break;
          case 'csv':
            console.log(versionsToCsv(versions));
            break;
          case 'table':
          default:
            printVersionsTable(versions);
            break;
        }
      } else {
        // Get work details
        const work = await getWork(workId);
        spinner.stop();

        switch (options.format.toLowerCase()) {
          case 'json':
            printJson(work);
            break;
          case 'csv':
            console.log('Note: CSV format not ideal for single work. Use table or json.');
            printWorkDetail(work);
            break;
          case 'table':
          default:
            printWorkDetail(work);
            break;
        }
      }
    } catch (error) {
      spinner.stop();
      if (error instanceof Error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
      }
      throw error;
    }
  });

File: nz-legislation/src/commands/get.rs
//! Get command implementation.

use clap::Args;
use crate::{config::Config, error::Result, output::{self, OutputFormat}};

/// Get legislation by ID.
#[derive(Debug, Args)]
pub struct CmdGet {
    /// Work ID (e.g., "act/2020/67" or "2020-67")
    #[arg(required = true)]
    pub id: String,

    /// Get version history
    #[arg(short, long)]
    pub versions: bool,
}

impl CmdGet {
    /// Execute the get command.
    pub async fn execute(
        &self,
        api_key: &str,
        base_url: &str,
        format: &OutputFormat,
    ) -> Result<()> {
        // Load configuration
        let config = Config::load_with_overrides(Some(api_key.to_string()), Some(base_url.to_string()))?;

        // Create API client
        let client = crate::client::ApiClient::new(config)?;

        if self.versions {
            // Get version history
            let versions = client.get_versions(&self.id).await?;
            
            match format {
                OutputFormat::Table => output::print_version_table(&versions),
                OutputFormat::Json => output::print_json(&versions)?,
                OutputFormat::Csv => output::print_version_csv(&versions)?,
            }
        } else {
            // Get work details
            let work = client.get_work(&self.id).await?;

            // Output results
            match format {
                OutputFormat::Table => output::print_work_detail(&work),
                OutputFormat::Json => output::print_json(&work)?,
                OutputFormat::Csv => {
                    // For single item, create a search results wrapper
                    let results = crate::models::SearchResults {
                        total: 1,
                        offset: 0,
                        limit: 1,
                        results: vec![work],
                        links: None,
                    };
                    output::print_work_csv(&results)?;
                }
            }
        }

        Ok(())
    }
}
